src/Razor/src/Microsoft.VisualStudio.Editor.Razor/Snippets/XmlSnippetParser.CodeSnippet.cs

```diff
@@ -0,0 +1,112 @@
+            }
+
+            CodeSnippetElement = codeSnippetElement;
+
+            Title = GetElementInnerText(header, "Title");
+            Shortcut = GetElementInnerText(header, "Shortcut");
+            var snippetTypes = GetElementsWithoutNamespace(header, "SnippetTypes");
+            if (snippetTypes != null)
+            {
+                _snippetTypes = snippetTypes.Elements().Select(e => e.Value.Trim()).ToArray();
+            }
+        }
+
+        public string Title { get; }
+
+        public string Shortcut { get; }
+
+        public XElement CodeSnippetElement { get; }
+
+        public bool IsExpansionSnippet()
+        {
+            return _snippetTypes?.Contains(ExpansionSnippetType, StringComparer.OrdinalIgnoreCase) == true;
+        }
+
+        public static CodeSnippet ReadSnippetFromFile(string filePath, string snippetTitle)
+        {
+            var document = XDocument.Load(filePath);
+            var snippets = ReadSnippets(document);
+            snippets.AssumeNotNull();
+
+            var matchingSnippet = snippets.Value.Single(s => string.Equals(s.Title, snippetTitle, StringComparison.OrdinalIgnoreCase));
+            return matchingSnippet;
+        }
+
+        private static ImmutableArray<XElement>? ReadCodeSnippetElements(XDocument document)
+        {
+            var codeSnippetsElement = document.Root;
+            if (codeSnippetsElement is null)
+                return null;
+
+            if (codeSnippetsElement.Name.LocalName.Equals("CodeSnippets", StringComparison.OrdinalIgnoreCase))
+            {
+                return codeSnippetsElement.Elements().Where(e => e.Name.LocalName.Equals("CodeSnippet", StringComparison.OrdinalIgnoreCase)).ToImmutableArray();
+            }
+            else if (codeSnippetsElement.Name.LocalName.Equals("CodeSnippet", StringComparison.OrdinalIgnoreCase))
+            {
+                return ImmutableArray.Create(codeSnippetsElement);
+            }
+
+            return null;
+        }
+
+        public static IEnumerable<XElement> GetElementsWithoutNamespace(XElement element, string localName)
+        {
+            return element.Elements().Where(e => e.Name.LocalName.Equals(localName, StringComparison.OrdinalIgnoreCase));
+        }
+
+        public static XElement? GetElementWithoutNamespace(XElement? element, string localName)
+        {
+            return element?.Elements().FirstOrDefault(e => e.Name.LocalName.Equals(localName, StringComparison.OrdinalIgnoreCase));
+        }
+
+        public static string GetElementInnerText(XElement element, string subElementName)
+        {
+            var subElement = GetElementWithoutNamespace(element, subElementName);
+            return subElement == null ? string.Empty : subElement.Value.Trim();
+        }
+
+        /// <summary>
+        /// Visible for testing.
+        /// </summary>
+        internal static ImmutableArray<CodeSnippet>? ReadSnippets(XDocument document)
```

Comment:
Consider returning `ImmutableArray<CodeSnippet>` rather than a nullable `ImmutableArray`. If you the other suggestions I've made for this type, it should result in simpler calling code.